posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,380 @@
+from posthog.models import Entity, Team
+from posthog.models import RetentionFilter
+from posthog.models.action.util import Action
+from posthog.models.filters.mixins.retention import RetentionDateDerivedMixin
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.team import WeekStartDay
+from posthog.queries.retention.types import CohortKey
+from posthog.queries.util import correct_result_for_sampling
+from posthog.queries.util import get_trunc_func_ch
+from posthog.schema import (
+    HogQLQueryModifiers,
+    RetentionQueryResponse,
+    IntervalType,
+)
+from posthog.schema import RetentionQuery, RetentionType
+
+
+class RetentionQueryRunner(QueryRunner):
+    query: RetentionQuery
+    query_type = RetentionQuery
+    EVENT_TABLE_ALIAS = "e"
+
+    def __init__(
+        self,
+        query: RetentionQuery | Dict[str, Any],
+        team: Team,
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        in_export_context: Optional[bool] = None,
+        event_query_type: Optional[RetentionQueryType] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, in_export_context=in_export_context)
+        self.event_query_type = event_query_type
+        self.old_filter = RetentionFilter(data=self.query.retentionFilter.model_dump())
+
+    def get_start_of_interval_sql(
+        self,
+        *,
+        source: str = "timestamp",
+        ensure_datetime: bool = False,
+    ) -> str:
+        trunc_func = get_trunc_func_ch(self.query.retentionFilter.period.name.lower())
+        trunc_func_args = [source]
+        if trunc_func == "toStartOfWeek":
+            trunc_func_args.append((WeekStartDay(self.team.week_start_day or 0)).clickhouse_mode)
+        interval_sql = f"{trunc_func}({', '.join(trunc_func_args)})"
+        return interval_sql
+
+    def retention_events_query(self):
+        _fields = [
+            self.get_timestamp_field(),
+            self.target_field(),
+        ]
+        params = {}
+
+        if self.event_query_type in [RetentionQueryType.TARGET, RetentionQueryType.TARGET_FIRST_TIME]:
+            source_timestamp = (
+                "min(e.timestamp)" if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME else "e.timestamp"
+            )
+
+            _fields += [
+                f"""
+                [
+                    dateDiff(
+                        {{period}},
+                        {self.get_start_of_interval_sql(source='{start_date}')},
+                        {self.get_start_of_interval_sql(source=source_timestamp)}
+                    )
+                ] as breakdown_values
+                """
+            ]
+
+        params.update(
+            {
+                "start_date": self.query_date_range.date_from(),
+                "period": self.query_date_range.interval_name,
+            }
+        )
```

Comment:
with things like `parse_expr`, you can still pass placeholders in if need be, which can avoid having these param objects floating about